### PR TITLE
Generate a correct consistent downsample target name in downsample ilm action

### PR DESCRIPTION
**Pull request by martijnvg**
Bring back generate downsample step that is used to prepare the lifecycle state and sets the name of the target downsample index to be used. A number of steps will use this to execute various downsample related operations.

The generate downsample index name step produces a consistent name. And also takes into account whether the source index was already downsampled, so a logical name is generated.

Relates to #93582

(marking as a non issue, because the bug isn't part of an released version of Elasticsearch)

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            newLifecycleState.build()
+        );
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    static String generateDownsampleIndexName(IndexMetadata sourceIndexMetadata, DateHistogramInterval fixedInterval) {
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java:58]: Hm, I'm not sure we need this persisted in the lifecycle execution state. 

Could we not derive this exact logic each time in `DownsampleStep`?

Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java:58]: Maybe I'm not understanding something correctly. If there is an easier way to do this then that would be great. But the issue here is that also the `downsampleAllocatedStep`, `copyExecutionStateStep`, `copyLifecycleSettingsStep`, `replaceDataStreamBackingIndex` and `swapAliasesAndDeleteSourceIndexStep` depend on the target name being generated. All step instances in the downsample action are shared, right? So it felt that the easiest way was to reuse the `LfecycleState#downsampleIndexName()` again.

Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java:58]: ++ this is what I was missing

So in theory we wouldn't need the persistence in the lifecycle state however the other steps are used across other actions and their refactoring would entail quite a bit of effort (as other actions do rely on index names coming from the lifecycle execution state as they are randomly generated)

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java [hunk 2]
```diff
@@ -314,6 +315,82 @@ public void testRollupNonTSIndex() throws Exception {
...
+            ),
+            30,
+            TimeUnit.SECONDS
+        );
+
+        // Manual rollover the original index such that it's not the write index in the data stream anymore
+        rolloverMaxOneDocCondition(client(), dataStream);
+
+        String downsampleIndexName = "downsample-" + firstBackingIndex + "-1m";
+        String downsampleOfDownsampleIndexName = "downsample-" + firstBackingIndex + "-1h";
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java:370]: Thanks for adding this test ❤️

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.LifecycleExecutionState;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
+
+import static org.elasticsearch.xpack.core.ilm.DownsampleAction.DOWNSAMPLED_INDEX_PREFIX;
+
+public class DownsamplePrepareLifeCycleStateStep extends ClusterStateActionStep {
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java:20]: Should we add a javadoc ?

**Review by andreidan — APPROVED** [review]
LGTM, thanks for fixing this Martijn

**Comment by martijnvg** [comment]
@elasticmachine update branch

**Comment by DaveCTurner** [comment]
Sorry, merging this PR introduced some compile errors so I have reverted it in 504aa401af723fb340a4abbcc8afd43130538c96. Please try again.

**Pull request merged**